app/service/s2s-service.ts

```diff
@@ -0,0 +1,102 @@
+import axios from 'axios';
+import { isJWTExpired } from '../utils/jwt-utils';
+import Logger from '../utils/logger';
+
+const config = require('config');
+const otp = require('otp');
+const s2sSecret = config.get('s2s.secret');
+const s2sUrl = config.get('s2s.url');
+const proxyHost = config.get('proxy.host');
+const proxyPort = config.get('proxy.port');
+const microServiceName = config.get('s2s.microserviceName');
+
+const logger: Logger = new Logger();
+const label = 's2s-service.ts';
+let serviceToken: string;
+
+interface IS2SService {
+  setServiceToken: (token: string) => void;
+  buildRequest: () => {};
+  requestServiceToken: () => void;
+  getServiceToken: () => {};
+}
+
+export default class S2SService implements IS2SService {
+  private static instance: S2SService;
+  private initialization;
+
+  public static getInstance(): S2SService {
+    if (!S2SService.instance) {
+      S2SService.instance = new S2SService();
+    }
+    return S2SService.instance;
+  }
+
+  constructor() {
+    this.initialization = this.init();
+  }
+
+  async init() {
+    await this.requestServiceToken();
+  }
+
+  /**
+   * Assembles a serviceAuthProvider request object to be used to query the service
+   * also creates a one-time-password from the secret.
+   */
+  buildRequest() {
+
+    const uri = `${s2sUrl}/lease`;
+    const oneTimePassword = otp(s2sSecret).totp();
+
+    return {
+      uri: uri,
+      body: {
+        microservice: microServiceName,
+        oneTimePassword: oneTimePassword
+      }
+    };
+  }
+
+  /**
+   * Sends out a request to the serviceAuthProvider and request a new service token
+   * to be passed as a header in any outgoing calls.
+   * Note: This token is stored in memory and this token is only valid for 3 hours.
+   */
+  async requestServiceToken() {
+    logger.trace('Attempting to request a S2S token', label);
+    const request = this.buildRequest();
+    let proxyConfig;
+    if (process.env.NODE_ENV === 'development') {
+      proxyConfig = { proxy: { host: proxyHost, port: proxyPort } };
+    }
+    let res;
+    try {
+      res = await axios.post(request.uri, request.body, proxyConfig);
+    } catch (err) {
+      logger.exception(err, label);
+    }
+    if (res && res.data) {
+      serviceToken = res.data;
+      logger.trace('Received S2S token and stored token', label);
+    } else {
+      logger.exception('Could not retrieve S2S token', label);
+    }
+  }
+
+  /**
+   * Checks that the token hasn't yet expired and retrieves the token that was saved in memory.
+   * If it has expired it refreshes the token and returns it.
+   */
+  async getServiceToken() {
+    if (isJWTExpired(serviceToken)) {
+      logger.trace('Token expired Attempting to acquire a new one.', label);
+      await this.requestServiceToken();
+    }
+    return serviceToken;
+  }
+
+  setServiceToken(token: string) {
+    serviceToken = token;
```

Comment:
if serviceToken is a private class attribute we could do this.serviceToken = token instead?